src/python/pants/engine/README.md

```diff
@@ -76,52 +76,72 @@ In cases where this search detects any ambiguity (generally because there are tw
+1. a compact and high-performance representation which can be stably shared across FFI boundaries,
+2. static type checking via [the `dataclasses` mypy plugin](https://github.com/python/mypy/blob/master/mypy/plugins/dataclasses.py),
+3. a concise, standard, and Pythonic way to declare classes.
+
+
+#### Example Usage of `@dataclass` for Engine `Param`s
+
+*Note that the [3rdparty `dataclasses` library](https://github.com/ericvsmith/dataclasses) must be in your BUILD file and `import`ed if you're running Pants with Python 3.6!*
 
 ```python
-class FormattedInt(datatype(['content'])): pass
+from dataclasses import dataclass
 
-@rule
-def int_to_str(value: int) -> FormattedInt:
-  return FormattedInt('{}'.format(value))
+from pants.util.meta import frozen_after_init
 
-# Field values can be specified with positional and/or keyword arguments in the constructor:
-x = FormattedInt('a string')
-x = FormattedInt(content='a string')
+# Pants requires that engine Params have a stable hash. This can be accomplished with the
+# `frozen=True` argument set in the `@dataclass` decorator.
+@dataclass(frozen=True)
+class FormattedInt:
+  content: str
 
-# Field values can be accessed after construction by name or index:
-print(x.content)    # 'a string'
-print(x[0])         # 'a string'
+# `@frozen_after_init` can also be used with `unsafe_hash=True` to create engine Params which can
+# modify their fields within the `__init__()` method:
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class ValidatedInt:
+  x: int
 
-# datatype objects can be easily inspected:
-print(x)            # 'FormattedInt(content=a string)'
-```
+  def __init__(self, x):
+    self.x = x
 
-#### Types of Fields
+@rule
+def int_to_str(value: ValidatedInt) -> FormattedInt:
+  return FormattedInt('{}'.format(value.x))
 
-`datatype()` accepts a list of *field declarations*, and returns a type which can be subclassed. A
-*field declaration* can just be a string (e.g. `'field_name'`), which is then used as the field
-name, as with `FormattedInt` above. A field can also be declared with a tuple of two elements: the
-field name string, and a `TypeConstraint` for the field (e.g. `('field_name',
-Exactly(FieldType))`). The bare type name (e.g. `FieldType`) can also be used as a shorthand for
-`Exactly(FieldType)`. If the tuple form is used, the constructor will create your object, then raise
-an error if the field value does not satisfy the type constraint.
+# Instances can be constructed and accessed the same way as for normal `@dataclass`es.
+x = FormattedInt('a string')
+assert x == FormattedInt(content='a string')
+print(x.content)                # 'a string'
 
-``` python
-class TypedDatatype(datatype([('field_name', Exactly(str, int))])):
+# `@dataclass` objects can be easily inspected:
+print(x)                        # 'FormattedInt(content="a string")'
+
+# All mypy types, including parameterized types, may be used in field definitions.
```

Comment:
```suggestion
# All MyPy types, including parameterized types, may be used in field definitions.
```

Well, true, but there are types we must avoid. For example, `List`, `Set`, and even `FrozenSet` are not safe (this one is up for discussion if we should make it work properly).

Maybe add a short section `### Types to avoid`?

Reply:
I ended up just emphasizing that the runtime type still has to be hashable, which would allow `FrozenSet` but not `List` or `Set`!